python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java

```diff
@@ -106,33 +107,66 @@ public static String fullyQualifiedModuleName(String packageName, String fileNam
     return globalSymbols;
   }
 
-  static void resolveTypeHierarchy(ClassDef classDef, @Nullable Symbol symbol, Map<Symbol, Set<Symbol>> subtypingRelations) {
+  static void resolveTypeHierarchy(ClassDef classDef, @Nullable Symbol symbol, PythonFile pythonFile, Map<String, Symbol> symbolsByName) {
     if (symbol == null || !Symbol.Kind.CLASS.equals(symbol.kind())) {
       return;
     }
     ClassSymbolImpl classSymbol = (ClassSymbolImpl) symbol;
-    Set<Symbol> superClasses = subtypingRelations.get(classSymbol);
-    if (superClasses != null) {
-      superClasses.forEach(classSymbol::addSuperClass);
+    if (isBuiltinTypeshedFile(pythonFile) && "str".equals(classSymbol.fullyQualifiedName())) {
+      classSymbol.addSuperClass(symbolsByName.get("object"));
+      classSymbol.addSuperClass(symbolsByName.get("Sequence"));
       return;
     }
     ArgList argList = classDef.args();
     if (argList == null) {
       return;
     }
     for (Argument argument : argList.arguments()) {
-      if (argument.is(Kind.REGULAR_ARGUMENT) && (((RegularArgument) argument).expression() instanceof HasSymbol)) {
-        Expression expression = ((RegularArgument) argument).expression();
-        Symbol argumentSymbol = ((HasSymbol) expression).symbol();
-        if (argumentSymbol != null) {
-          classSymbol.addSuperClass(argumentSymbol);
-          continue;
+      Symbol parentSymbol = getParentSymbol(argument);
+      if (parentSymbol == null) {
+        classSymbol.setHasSuperClassWithoutSymbol();
+      } else {
+        Symbol resolvedParentSymbol = resolveParentSymbol(parentSymbol, pythonFile, symbolsByName);
+        if (resolvedParentSymbol != null) {
+          classSymbol.addSuperClass(resolvedParentSymbol);
         }
       }
-      classSymbol.setHasSuperClassWithoutSymbol();
     }
   }
 
+  private static Symbol resolveParentSymbol(Symbol parentSymbol, PythonFile pythonFile, Map<String, Symbol> symbolsByName) {
+    if (isTypingFile(pythonFile) && (parentSymbol.name().equals("Protocol") || parentSymbol.name().equals("Generic"))) {
+      // ignore Protocol and Generic to avoid having incomplete type hierarchies
+      return null;
+    }
+    if (isTypingFile(pythonFile) && parentSymbol.name().equals("_Collection")) {
+      return symbolsByName.get("Collection");
+    }
+    return parentSymbol;
+  }
+
+  private static boolean isBuiltinTypeshedFile(PythonFile pythonFile) {
+    return isTypeShedFile(pythonFile) && pythonFile.fileName().isEmpty();
+  }
+
+  private static boolean isTypingFile(PythonFile pythonFile) {
+    return isTypeShedFile(pythonFile) && pythonFile.fileName().equals("typing");
+  }
+
+  @CheckForNull
+  private static Symbol getParentSymbol(Argument argument) {
```

Comment:
I don't understand what we mean by "parent" here.

Reply:
"parent" is refererring to "parent class (of the current class)". I understand the confusion, I'll rename it "getSymbolFromArgument"

Reply:
Fixed in 85f07023b31df792b45e866765cbcc6eb4e45c09